joblib/_parallel_backends.py

```diff
@@ -320,12 +356,38 @@ def configure(self, n_jobs=1, parallel=None, **backend_args):
     def _get_pool(self):
         """Lazily initialize the thread pool
 
-        The actual pool of worker threads is only initialized at the first
+        The actual pool of worker threads is shared across
+        ThreadingBackends in a process and only initialized at the first
         call to apply_async.
         """
-        if self._pool is None:
-            self._pool = ThreadPool(self._n_jobs)
-        return self._pool
+        global _thread_pool
+        global _thread_pool_users
+        if _thread_pool is None:
+            # We need to create a thread pool
+            _thread_pool = SafeThreadPool(self._n_jobs)
+            # We own it. It is our responsability to close it
+            self._pool = _thread_pool
+        else:
+            # Try to get more jobs
+            # The code below is accessing multiprocessing private API
+            max_processes = cpu_count() + len(_thread_pool_users)
+            if _thread_pool._processes < max_processes:
+                _thread_pool._processes = min(max_processes,
+                    _thread_pool._processes + self._n_jobs)
+                _thread_pool.repopulate_pool()
+        _thread_pool_users.add(self)
+        return _thread_pool
+
+    def terminate(self):
+        """Shutdown the process or thread pool"""
+        if self._pool is not None:
+            super(ThreadingBackend, self).terminate()
+            global _thread_pool
+            _thread_pool = None
+        try:
+            _thread_pool_users.remove(self)
```

Comment:
I think you are missing a `global` here?

Also, I don't understand why you terminate the `ThreadPool` if `len(_thread_pool_user) > 1`.  In this case, you should just remove some thread from it no?

Reply:
I don't think the global declaration is required: we are just mutating the value of the _thread_pool_users variable, not assigning it a new value.